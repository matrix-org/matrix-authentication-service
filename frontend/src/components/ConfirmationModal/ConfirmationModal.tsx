// Copyright 2023, 2024 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Button } from "@vector-im/compound-web";
import { ReactNode } from "react";
import { Translation } from "react-i18next";

import { Dialog, Title, Description, Close, Actions } from "../Dialog";

type Props = {
  onConfirm: () => void;
  title?: ReactNode | string;
  // element used to trigger opening of modal
  trigger: ReactNode;
};

/**
 * Generic confirmation modal
 * controls its own open state
 * calls onConfirm on confirm click
 */
const ConfirmationModal: React.FC<React.PropsWithChildren<Props>> = ({
  onConfirm,
  children,
  trigger,
  title,
}) => (
  <Translation>
    {(t): ReactNode => (
      <Dialog trigger={trigger}>
        <Title>{title}</Title>

        {children && <Description>{children}</Description>}

        <Actions>
          <Close asChild>
            <Button kind="primary" destructive onClick={onConfirm}>
              {t("action.continue")}
            </Button>
          </Close>

          <Close asChild>
            <Button kind="tertiary">{t("action.cancel")}</Button>
          </Close>
        </Actions>
      </Dialog>
    )}
  </Translation>
);

export default ConfirmationModal;
