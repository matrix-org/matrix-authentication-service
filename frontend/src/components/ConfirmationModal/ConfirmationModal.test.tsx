// Copyright 2023 The Matrix.org Foundation C.I.C.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// @vitest-environment happy-dom

import { render, cleanup, screen, fireEvent } from "@testing-library/react";
import { describe, expect, it, afterEach, vi } from "vitest";

import ConfirmationModal from "./ConfirmationModal";

describe("<ConfirmationModal />", () => {
  afterEach(cleanup);

  const trigger = <button>Open modal</button>;

  it("does not render a closed modal", () => {
    const onConfirm = vi.fn();
    const onDeny = vi.fn();

    render(
      <ConfirmationModal
        onConfirm={onConfirm}
        onDeny={onDeny}
        className="test"
        trigger={trigger}
        title="Are you sure?"
      >
        Some extra information.
      </ConfirmationModal>,
    );
    expect(screen.getByText("Open modal")).toBeTruthy();
    expect(screen.queryByRole("alertdialog")).toBeFalsy();
  });

  it("opens modal on clicking trigger", () => {
    const onConfirm = vi.fn();
    const onDeny = vi.fn();

    render(
      <ConfirmationModal
        onConfirm={onConfirm}
        onDeny={onDeny}
        className="test"
        trigger={trigger}
        title="Are you sure?"
      >
        Some extra information.
      </ConfirmationModal>,
    );

    fireEvent.click(screen.getByText("Open modal"));

    expect(screen.getByRole("alertdialog")).toMatchSnapshot();
  });

  it("renders an undeniable modal", () => {
    const onConfirm = vi.fn();
    const onDeny = undefined;

    render(
      <ConfirmationModal
        onConfirm={onConfirm}
        onDeny={onDeny}
        className="test"
        trigger={trigger}
        title="Are you sure?"
      >
        Some extra information.
      </ConfirmationModal>,
    );

    fireEvent.click(screen.getByText("Open modal"));

    // no cancel button without onDeny
    expect(screen.queryByText("Cancel")).toBeFalsy();

    // The dialog does not close on escape
    fireEvent.keyDown(screen.getByRole("alertdialog"), {
      key: "Escape",
      code: "Escape",
      keyCode: 27,
    });

    // dialog still open
    expect(screen.queryByRole("alertdialog")).toBeTruthy();
  });

  it("calls onConfirm on confirmation", () => {
    const onConfirm = vi.fn();
    const onDeny = vi.fn();

    render(
      <ConfirmationModal
        onConfirm={onConfirm}
        onDeny={onDeny}
        className="test"
        trigger={trigger}
        title="Are you sure?"
      >
        Some extra information.
      </ConfirmationModal>,
    );

    fireEvent.click(screen.getByText("Open modal"));

    fireEvent.click(screen.getByText("Continue"));
    expect(onConfirm).toHaveBeenCalled();

    // dialog closed
    expect(screen.queryByRole("alertdialog")).toBeFalsy();
  });

  it("calls onDeny on cancel click", () => {
    const onConfirm = vi.fn();
    const onDeny = vi.fn();

    render(
      <ConfirmationModal
        onConfirm={onConfirm}
        onDeny={onDeny}
        className="test"
        trigger={trigger}
        title="Are you sure?"
      >
        Some extra information.
      </ConfirmationModal>,
    );

    fireEvent.click(screen.getByText("Open modal"));

    fireEvent.click(screen.getByText("Cancel"));
    expect(onDeny).toHaveBeenCalled();

    // dialog closed
    expect(screen.queryByRole("alertdialog")).toBeFalsy();
  });

  it("calls onDeny on closing modal via Esc", () => {
    const onConfirm = vi.fn();
    const onDeny = vi.fn();

    render(
      <ConfirmationModal
        onConfirm={onConfirm}
        onDeny={onDeny}
        className="test"
        trigger={trigger}
        title="Are you sure?"
      >
        Some extra information.
      </ConfirmationModal>,
    );

    fireEvent.click(screen.getByText("Open modal"));

    fireEvent.keyDown(screen.getByRole("alertdialog"), {
      key: "Escape",
      code: "Escape",
      keyCode: 27,
    });
    expect(onDeny).toHaveBeenCalled();

    // dialog closed
    expect(screen.queryByRole("alertdialog")).toBeFalsy();
  });
});
