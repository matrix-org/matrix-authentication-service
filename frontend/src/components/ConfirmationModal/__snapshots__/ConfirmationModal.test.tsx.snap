// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ConfirmationModal /> > opens modal on clicking trigger 1`] = `
<div
  aria-describedby="radix-:r5:"
  aria-labelledby="radix-:r4:"
  class="_content_6bc2cb test"
  data-state="open"
  id="radix-:r3:"
  role="alertdialog"
  style="pointer-events: auto;"
  tabindex="-1"
>
  <h2
    id="radix-:r4:"
  >
    Are you sure?
  </h2>
  <p
    id="radix-:r5:"
  >
    Some extra information.
  </p>
  <div
    class="_buttons_6bc2cb"
  >
    <button
      class="_button_1ahdh_17"
      data-kind="tertiary"
      data-size="sm"
      role="button"
      tabindex="0"
      type="button"
    >
      Cancel
    </button>
    <button
      class="_button_1ahdh_17"
      data-kind="destructive"
      data-size="sm"
      role="button"
      tabindex="0"
      type="button"
    >
      Continue
    </button>
  </div>
</div>
`;
